### PR TITLE
Convert to Yarn workspace

### PR DESCRIPTION
**Pull request by marcoow**
As we're starting to make substantial modifications to the codebase that will also significantly change the way that ESA is used in applications (and likely drop support for some older versions of Ember again), we will require significantly different applications to test ESA in (e.g. one using the classic mixins integration, one using ESA via decorators (or another new mechanism), one app using typescript maybe etc.) This PR creates the foundation for that so that we can have as many Ember apps as we want as workspaces that all use the same addon.

### TODO

- [ ] the Procfiles are now both at the top level and called `Procfile.client` and `Procfile.server` – we'll need to update the heroku settings accordingly when merging this

### CODE REVIEW

File: packages/ember-simple-auth/node_modules/supports-color/package.json [hunk 1]
```diff
@@ -0,0 +1,53 @@
+{
```
Comment (Turbo87) [inline packages/ember-simple-auth/node_modules/supports-color/package.json:1]: why do we add `node_modules` folders to this repo?

Comment (marcoow) [reply packages/ember-simple-auth/node_modules/supports-color/package.json:1]: Ugh, added that accidentally.

This is still an early draft though – I'll let everyone know when this is ready for review.

**Comment by marcoow** [comment]
It's likely I missed some things and unfortunately this PR is huge but it'd be good to get a second (and third, fourth, …) pair of eyes on this.

File: packages/ember-simple-auth/node_modules/.bin/_mocha [hunk 2]
```diff
@@ -0,0 +1 @@
+../mocha/bin/_mocha
```
Comment (pichfl) [inline packages/ember-simple-auth/node_modules/.bin/_mocha:1]: Are these checked in intentionally?

**Comment by pichfl** [comment]
@marcoow I think some `packages/ember-simple-auth/node_modules/*` were added accidentally

File: packages/ember-simple-auth/node_modules/.bin/_mocha (hunk 2 above)
Comment (marcoow) [reply packages/ember-simple-auth/node_modules/.bin/_mocha]: Ugh yep, added those accidentally – removed now

File: .ember-cli [hunk 3]
```diff
@@ -1,9 +0,0 @@
-{
-  /**
-    Ember CLI sends analytics information by default. The data is completely
-    anonymous, but there are times when you might want to disable this behavior.
-
-    Setting `disableAnalytics` to true will prevent any data from being sent.
-  */
-  "disableAnalytics": false
-}
```
Comment (Turbo87) [inline .ember-cli:9]: I'm surprised that this doesn't show up as "moved". Did you explicitly delete it?

File: .eslintignore [hunk 4]
```diff
@@ -1,9 +1,3 @@
-/blueprints/*/files
-
-# unconventional js
-/blueprints/*/files/
-/vendor/
-
 # compiled output
 /dist/
 /tmp/
```
Comment (Turbo87) [inline .eslintignore:3]: why do we keep these? do we expect to have these folders in the top-level folder?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -35,22 +41,24 @@ jobs:
...
 
   tests:
     name: Tests
     runs-on: ubuntu-latest
 
     strategy:
       matrix:
+        workspace:
+          - ember-simple-auth
+          - test-app
```
Comment (Turbo87) [inline .github/workflows/ci.yml:54]: do we want to run these workspace tests for all Ember versions? I'm a bit worried the CI time will explode by doing that.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -35,22 +41,24 @@ jobs:
         run: yarn install
 
       - name: lint
-        run: yarn lint
+        run: yarn workspace ${{ matrix.workspace }} lint
```
Comment (Turbo87) [inline .github/workflows/ci.yml:44]: I'm wondering if we should run the lint for all workspaces in just one job instead

File: packages/ember-simple-auth/.eslintignore [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+/tmp/
+
+# dependencies
+/bower_components/
+
+# misc
+/coverage/
+
+# ember-try
+/.node_modules.ember-try/
```
Comment (Turbo87) [inline packages/ember-simple-auth/.eslintignore:18]: what about `.bower_components.ember-try`? otherwise we probably don't need the `bower_components` above either, right?

File: packages/ember-simple-auth/ember-cli-build.js [hunk 8]
```diff
@@ -9,11 +9,6 @@ var sourceTrees = [];
 module.exports = function(defaults) {
   var app = new EmberAddon(defaults, {
     storeConfigInMeta: true,
-    'ember-bootstrap': {
-      bootstrapVersion: 4,
-      importBootstrapFont: false,
-      importBootstrapCSS: true
-    },
```
Comment (Turbo87) [inline packages/ember-simple-auth/ember-cli-build.js:16]: why was this removed?

File: packages/ember-simple-auth/tests/unit/mixins/application-route-mixin-test.js [hunk 9]
```diff
@@ -29,6 +31,7 @@ describe('ApplicationRouteMixin', () => {
       clear: sinon.stub()
     }));
 
+    this.owner.register('route:application', Route.extend(ApplicationRouteMixin));
```
Comment (Turbo87) [inline packages/ember-simple-auth/tests/unit/mixins/application-route-mixin-test.js:34]: why is this change necessary now?

File: .ember-cli (hunk 3 above)
Comment (marcoow) [reply .ember-cli]: fixed

File: .eslintignore (hunk 4 above)
Comment (marcoow) [reply .eslintignore]: fixed

File: .github/workflows/ci.yml (hunk 5 above)
Comment (marcoow) [reply .github/workflows/ci.yml:54]: I think we need to do this, otherwise we lose a bit of coverage as the tests in `packages/ember-simple-auth` are unit tests only and the ones in `packages/test-app` are application tests.

File: .github/workflows/ci.yml (hunk 6 above)
Comment (marcoow) [reply .github/workflows/ci.yml:44]: But this way we can run the jobs in parallel?

File: packages/ember-simple-auth/ember-cli-build.js (hunk 8 above)
Comment (marcoow) [reply packages/ember-simple-auth/ember-cli-build.js:16]: It is no longer needed in `packages/ember-simple-auth` since the dummy app has been moved to `packages/test-app` and the dummy app in `packages/ember-simple-auth` is now empty.

File: packages/ember-simple-auth/.eslintignore (hunk 7 above)
Comment (marcoow) [reply packages/ember-simple-auth/.eslintignore:15]: I removed `bower_components` above 👍

File: packages/ember-simple-auth/tests/unit/mixins/application-route-mixin-test.js (hunk 9 above)
Comment (marcoow) [reply packages/ember-simple-auth/tests/unit/mixins/application-route-mixin-test.js:34]: This was previously relying on the application route that's defined in the dummy app. The dummy app of `packages/ember-simple-auth` is now empty though as it has been moved to `packages/test-app`, thus there is no application controller anymore if we don't register one.

File: .github/workflows/ci.yml (hunk 6 above)
Comment (Turbo87) [reply .github/workflows/ci.yml:44]: yeah, but it has to perform the `yarn install` for every single job which most likely takes longer than the actual linting

Comment (marcoow) [reply .github/workflows/ci.yml:44]: The linting jobs still finish in around 1m though: https://github.com/simplabs/ember-simple-auth/pull/2184/checks?check_run_id=608604147 which is substantially faster than the jobs than run tests. Assuming all jobs run in parallel I think keeping the CI config simpler trumps speeding up these jobs that are not a limiting factor for the CI pipeline anyway?

File: package.json [hunk 10]
```diff
@@ -1,124 +1,15 @@
...
-    "configPath": "tests/dummy/config",
-    "versionCompatibility": {
-      "ember": ">=1.12"
-    }
-  },
-  "fastbootDependencies": [
-    "node-fetch"
+  "workspaces": [
+    "packages/*"
   ],
```
[the remark this answers is not in the document]
Comment (marcoow) [reply package.json:10]: What would you say would the advantage of that be?

File: package.json [hunk 11]
```diff
@@ -1,124 +1,15 @@
...
-    "loader.js": "^4.5.1",
-    "marked": "^1.0.0",
-    "mermaid": "^7.0.3",
-    "mocha": "^7.0.1",
-    "morgan": "^1.10.0",
-    "my-engine": "./tests/dummy/lib/my-engine",
-    "rimraf": "^3.0.0",
-    "sinon-chai": "~3.5.0",
-    "torii": "^0.10.0"
+    "diagrams": "mmdc -i guides/assets/esa-initial-flow.txt -o guides/assets/esa-initial-flow.svg"
```
[the remark this answers is not in the document]
Comment (marcoow) [reply package.json:6]: Not sure what you mean – the chart is linked from the [guides](https://github.com/simplabs/ember-simple-auth/tree/master/guides)

File: .github/workflows/ci.yml [hunk 12]
```diff
@@ -105,5 +119,32 @@ jobs:
...
+
+      - uses: actions/cache@v1
+        with:
+          path: ${{ steps.yarn-cache.outputs.dir }}
+          key: ${{ runner.os }}-yarn-test-${{ hashFiles('**/yarn.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-yarn-
+
+      - name: install dependencies
+        run: yarn install
```
[the remark this answers is not in the document]
Comment (marcoow) [reply .github/workflows/ci.yml:147]: Yes, probably a good idea but unrelated as you say 👍

File: package.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (patsy-issa) [reply package.json:6]: Ah missed the link in there 👍

The previous script generates a `png`  that the guides link to, the following change turns the output to `svg`, I would prefer if we keep the original `png` output for this pr and replace the png resources with svgs on its own.

File: package.json (hunk 10 above)
[the remark this answers is not in the document]
Comment (patsy-issa) [reply package.json:10]: I think it makes it easier to see what workspaces exists in the project from `package.json` without needing to look under `/packages`

File: package.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (marcoow) [reply package.json:6]: This newer version of mermaid doesn't support PNG anymore I think.

File: package.json (hunk 10 above)
[the remark this answers is not in the document]
Comment (marcoow) [reply package.json:10]: Hm, we can list the packages of course – I'm not sure that's a real problem anyone would have though? 🤔

[the remark this answers is not in the document]
Comment (patsy-issa) [reply package.json:10]: Not a problem per say more of a explicit vs automagic preference, no strong feelings either way ^^

[the remark this answers is not in the document]
Comment (marcoow) [reply package.json:10]: I don't really see added value tbh – let's keep it as is if you don't have a strong preference

File: package.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (marcoow) [reply package.json:6]: fixed in the guides

File: guides/auth-torii-with-github.md [hunk 13]
```diff
@@ -31,7 +31,7 @@ configuration and base classes. For GitHub specifically, the authentication flow
    <access_token>` header request to, say, `https://api.github.com/user`,
    which returns the GitHub user’s user info, at which point it can be matched to the app's own `User` model. 
 
-You can find a sequence diagram for the full flow [here](./assets/esa-initial-flow.txt.png).
+You can find a sequence diagram for the full flow [here](./assets/esa-initial-flow.svg).
```
Comment (patsy-issa) [inline guides/auth-torii-with-github.md:34]: Won't this be a broken link as the [file](https://github.com/simplabs/ember-simple-auth/tree/23981b7e39d137bd63440bdc114bcf7daa832ef8/guides/assets) was not generated manually?

Comment (marcoow) [reply guides/auth-torii-with-github.md:34]: 🤦 fixed
